compute-io/add
# Update according to TODO

- copy option
- accessor option
- Makefile
- README
- validation dependencies
- dev dependencies
- remove preamble
- update license info
- tests for code pathways
- dotfiles

Resolves compute-io/todo#40

**NOTE**: breaking change. `add` no longer mutates by default. This behavior must be explicitly set using the `copy` option. While memory management is important, three factors contribute to turning **off** mutation by default:
1. _principle of least surprise_: mutation is a source of bugs, especially when multiple operations may operate on the same data store. Reasoning about mutable objects as they travel through a program with black boxes (functions mutating state) is more complex.
2. consideration that, for object `arrays`, mutation has a side-effect that object elements are eliminated and replaced with numeric results. For more complex object `arrays`, this is unlikely to be desired. Setting `{'copy':true}` for each invocation is probably more than should be asked.
3. readability. If mutation is desired, having the invocation signature `add( x, y, {'copy': false})` means that the code is explicit in describing the behavior. If I understand that the `copy` option pertains to mutation, then I derive meaning from the signature.
